scripts/run-l10n-extraction

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env bash
+
+# Exit immediately when a command fails.
+set -e
+# Make sure exit code are respected in a pipeline.
+set -o pipefail
+# Treat unset variables as an error and exit immediately.
+set -u
+
+INITIAL_GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+GIT_CHANGES=$(git status --porcelain)
+DIFF_WITH_ONE_LINE_CHANGE="1 file changed, 1 insertion(+), 1 deletion(-)"
+GIT_REPO_NAME=addons-linter
+GIT_REMOTE="${GIT_REMOTE:-git@github.com:mozilla/$GIT_REPO_NAME.git}"
+LOG_FILE="l10n-extraction.log"
+
+info() {
+  local message="$1"
+
+  echo ""
+  echo "INFO: $message"
+  echo ""
+}
+
+error() {
+  local message="$1"
+
+  echo "ERROR: $message"
+  exit 1
+}
+
+run_l10n_extraction() {
+  local branch="$GIT_REPO_NAME-locales"
+
+  info "Starting l10n extraction"
+
+  # Detect local (uncommitted) changes.
+  if [[ ! -z "$GIT_CHANGES" ]]; then
+    error "You have local changes, therefore this script cannot continue."
+  fi
+
+  # Switch to the `master` branch if we are not on it already.
+  if [[ "$INITIAL_GIT_BRANCH" != "master" ]]; then
+    git checkout master
+  fi
```

Comment:
I gave this script a try, given that I did have a PR with changes to the locales.

I think that I get why the script assumes that we always want to regenerate locales from the master branch and not arbitrary feature branches, but doing it automatically and automatically create a branch in the official upstream repo feels a bit too auto-magic to me and I'm wondering if that may be surprising only from my perspective. 

As an example, as in this case, as a developer my intention may be to fix a string and to ensure the same PR does also update the locale files accordingly, but that isn't really the case, it is not actually how the script is meant to be used.

Not a blocking issue (especially given that I assume we are using and heppy with this strategy on the addons-frontend side for quite some time), but I'm wondering if:

- when the current branch isn't the master branch, then failing explicitly and printing a guidance message (that explains this is meant to only run on a clean master branch) may make this a bit less surprising
  - or alternatively to ask confirmation first (after printing a message to explain this is only meant to regenerate locales on the master branch

- deleting a branch with `-D` is slightly scary, I see that there shouldn't be too many chances to remove the wrong branch in the end, but asking for confirmation may still be reasonable (in the end this doesn't seem a script we intend to run on CI and I guess we may not need to run it too ofte, I guess ~1-2 times during a cycle and before each release may be enough)

- Pushing to the upstream remote repo is also something I'm wondering if we could ask a confirmation before doing it, maybe even after having printed the `git log --stat` output for the new commits range 

@willdurand wdyt?

Reply:
> I gave this script a try, given that I did have a PR with changes to the locales.
> 
> I think that I get why the script assumes that we always want to regenerate locales from the master branch and not arbitrary feature branches, but doing it automatically and automatically create a branch in the official upstream repo feels a bit too auto-magic to me and I'm wondering if that may be surprising only from my perspective.

We created this script in addons-frontend to remove friction because the old process was manual with too many steps. This script automates all these steps, without any confirmation because that would be equally annoying.

> As an example, as in this case, as a developer my intention may be to fix a string and to ensure the same PR does also update the locale files accordingly, but that isn't really the case, it is not actually how the script is meant to be used.

Yeah, that isn't possible. Except for #4024, we usually add or update new strings but we don't do the l10n extraction as part of the PR. We usually do the l10n extraction from time to time (although with a script, I would probably do more frequently). In addons-frontend, l10n extraction happens every week after the code freeze.

> Not a blocking issue (especially given that I assume we are using and heppy with this strategy on the addons-frontend side for quite some time), but I'm wondering if:
> 
>     * when the current branch isn't the master branch, then failing explicitly and printing a guidance message (that explains this is meant to only run on a clean master branch) may make this a bit less surprising

It is already aborting when the repo is in a dirty state. I don't think we should abort because the process *requires* to do the l10n extraction on the master branch. We could improve the script to by pass this check, though, which would cover the use case mentioned above (but not sure it is worth it).

> or alternatively to ask confirmation first (after printing a message to explain this is only meant to regenerate locales on the master branch

Maybe? We didn't have docs about the l10n extraction process but the addons-frontend instructions already mention that: https://addons-frontend.readthedocs.io/en/latest/i18n/#updating-locales. I think that's why it isn't a surprise for me or the rest of the AMO team.

> deleting a branch with `-D` is slightly scary, I see that there shouldn't be too many chances to remove the wrong branch in the end, but asking for confirmation may still be reasonable

I could rename the conflicting branch or use a longer branch name, potentially including the datetime?

> (in the end this doesn't seem a script we intend to run on CI 

I'd be good to fully automate this process with CI but... we never managed to do that on other AMO project.

> I guess we may not need to run it too ofte, I guess ~1-2 times during a cycle and before each release may be enough)

I think it'd be nice to run this script after every release.

>  Pushing to the upstream remote repo is also something I'm wondering if we could ask a confirmation before doing it, maybe even after having printed the `git log --stat` output for the new commits range

Isn't the summary printed at the end enough? It is going to push a feature branch and only users with WRITE access can do that anyway.

---

To summarize, the "no-confirmation" approach is a feature, not a bug. I can make changes to avoid branch conflicts and maybe some more `info` messages but I am not convinced that we should ask confirmation.

(FWIW the script in addons-frontend has been used by 4 devs on 3 different OSes every week for the last 18 months without much problems. There were some issues with Windows initially but no work has been lost).

Reply:
I suppose the README section should make it crystal-clear that the script runs without any confirmation and explain the different main steps?